Add Swift Package Manager docs

loic-sharma opened the pull request:
Flutter iOS/macOS is migrating to Swift Package Manager for native dependencies. This is an experimental feature this isn't ready yet for app developers. However, plugin authors should add Swift Package Manager support to their packages.

The faster plugin authors add SwiftPM support, the sooner we can drop CocoaPods (and make it easier to install Flutter iOS!). There's [~10K Flutter iOS plugins](https://gist.github.com/loic-sharma/891a358f8abab36133e8a9395f719296) that'll need to migrate, so we really need the plugin migration docs to be as easy to follow as possible. Any and all feedback is appreciated! :)
 
This is based off @vashworth's [guide](https://github.com/flutter/flutter/pull/146256#issuecomment-2038565077).

Question for reviewers: What do you think of the doc structure? Should we rearrange things to make it clearer that some sections are to be followed when things go wrong? Or is the current structure clear enough?

Preview URLs:

1. [Swift Package Manager for app developers](https://flutter-docs-prod--pr10827-spm-compatibility-ilxd3b00.web.app/packages-and-plugins/swift-package-manager/for-app-developers)
2. [Swift Package Manager for plugin authors
](https://flutter-docs-prod--pr10827-spm-compatibility-ilxd3b00.web.app/packages-and-plugins/swift-package-manager/for-plugin-authors)

Part of https://github.com/flutter/flutter/issues/126005

## Presubmit checklist

- [ ] This PR is marked as draft with an explanation if not meant to land until a future stable release.
- [x] This PR doesn’t contain automatically generated corrections (Grammarly or similar).
- [x] This PR follows the [Google Developer Documentation Style Guidelines](https://developers.google.com/style) — for example, it doesn’t use _i.e._ or _e.g._, and it avoids _I_ and _we_ (first person).
- [x] This PR uses [semantic line breaks](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) of 80 characters or fewer.

hellohuanlin commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,62 @@
+Flutter's Swift Package Manager feature is disabled by default.
+To enable it:
+
+1. Switch to Flutter's `master` channel:
```
do we still use master channel or is it renamed to main?

loic-sharma replied:
Both exist but `master` is still the "real" one.

sethladd replied:
If `main` works, let's go with main.

loic-sharma replied:
Turns out I was wrong, switched to `main`!

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,62 @@
+Package Manager yet.
+:::
+
+## How to disable Swift Package Manager
```
Ideally app developers enable SwiftPM and never disable it.

However, plugin developers need to enable/disable SwiftPM to test their plugin works with both scenarios.

loic-sharma commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,121 @@
+
+[flutter#146957]: https://github.com/flutter/flutter/issues/146957
+
+## How to add Swift Package Manager integration to a project manually
```
Our hope is that the automatic migration will migrate most projects. Hopefully most users won't need to follow these steps.

However, users can modify their project in a way that breaks out automatic migration. Ideally users would let us know about these modifications so that we can improve our automatic migration.

Question for reviewers: Is it clear that most people can skip these steps?

atsansone replied:
**suggestion:** Remove _manually_. If the reader is doing something, then it's manual by definition.

```suggestion
## How to add Swift Package Manager integration to a project
```

loic-sharma replied:
I removed manually from the section header but clarified the section intro to clarify these steps are only necessary if the automatic migration fails.

loic-sharma commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,121 @@
+{% endtab %}
+{% endtabs %}
+
+## How to use a Swift Package Manager Flutter plugin that requires a higher OS version
```
This is a new kind of error that users can hit when using SwiftPM.

loic-sharma commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,121 @@
+image:"development/packages-and-plugins/swift-package-manager/minimum-deployments.png",
+caption:"The target's Minimum Deployments setting" %}
+
+## How to add Swift Package Manager integration to a custom target
```
This is for advanced scenarios where users have heavily modified their project. Most users won't need these steps. Please let me know if there's a better way to call out this special case without making the "manual migration" docs overly convoluted.

jmagman replied:
Maybe "to a custom Xcode target" and then explain what that means in the first sentence?

loic-sharma replied:
Added!

loic-sharma commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,121 @@
+
+In Step 2, list item 6 use your custom target instead of the Flutter target.
+
+## How to remove Swift Package Manager integration
```
In theory no one needs to do these steps unless there's a horrible bug in our SwiftPM integration.

sethladd commented on src/content/packages-and-plugins/swift-package-manager/for-plugin-authors.md:
```diff
@@ -0,0 +1,132 @@
+diff2html: true
+---
+
+Flutter is migrating to Swift Package Manager to manage iOS and macOS native
```
should we link to "Swift Package Manager" here, for moar better SEO and help those who aren't aware of it?

loic-sharma replied:
Linked!

sethladd commented on src/content/packages-and-plugins/swift-package-manager/for-plugin-authors.md:
```diff
@@ -0,0 +1,132 @@
+Flutter is migrating to Swift Package Manager to manage iOS and macOS native
+dependencies.
+This is an experimental feature that may change in the future.
+It is currently only available on the [`master` channel][].
```
consider: 'main'

loic-sharma replied:
Turns out I was wrong, switched to `main`!

sethladd commented on src/content/packages-and-plugins/swift-package-manager/for-plugin-authors.md:
```diff
@@ -0,0 +1,132 @@
+It is currently only available on the [`master` channel][].
+Flutter will continue to support CocoaPods until further notice.
+
+Swift Package Manager adoption will be gradual.
```
if we don't know how the qualify this, not sure we need to say this. I get the sense we want to migrate the ecosystem faster than slower (but, not recklessly)

loic-sharma replied:
Ah did you read this as: "Feel free to take your time, no rush on this migration"?

The intent behind this sentence was slightly different: plugin authors should **not** make plugins that only support SwiftPM. This would break users that haven't updated their projects yet. Plugin should support _both_ CocoaPods and SwiftPM until further notice.

I'll rephrase to make this clearer :)

loic-sharma replied:
I tweaked this intro, let me know if it's clearer now.

sethladd commented on src/content/packages-and-plugins/swift-package-manager/for-plugin-authors.md:
```diff
@@ -0,0 +1,132 @@
+Swift Package Manager adoption will be gradual.
+
+Flutter recommends that Flutter plugins support _both_ Swift Package Manager
+and CocoaPods until further notice.
```
if that's not easy to do, should we link to guidance for how to support both?

loic-sharma replied:
This guidance does make your plugin support both package managers. I'll add a blurb to clarify this.

loic-sharma replied:
I tweaked this intro, let me know if it's clearer now.

sethladd commented on src/content/packages-and-plugins/swift-package-manager/index.md:
```diff
@@ -0,0 +1,6 @@
+---
+layout: toc
+title: Swift Package Manager
```
maybe: "Swift Package Manager for Flutter"

loic-sharma replied:
Updated!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+However, the Flutter CLI may be unable to migrate your project if you've
+modified it significantly.
+
+Please [file a bug][] before migrating manually to help the Flutter team improve
+the automatic migration.
+Please include the error message you received and consider including a copy of
+the of the following files in your bug report:
```
```suggestion
If auto-migration fails, please [file an issue][]
before migrating manually; this helps the Flutter team improve
the automatic migration process.
Include the error message and, if possible, include a copy of
the following files in your issue:
```

loic-sharma replied:
Thanks!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+### Step 1: Add FlutterGeneratedPluginSwiftPackage Package Dependency {:.no_toc}
+
+1. Open your app (`ios/Runner.xcworkspace`) in Xcode.
+2. Navigate to Package Dependencies for the project.
```
```suggestion
2. Navigate to **Package Dependencies** for the project.
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+   image:"development/packages-and-plugins/swift-package-manager/package-dependencies.png",
+   caption:"The project's package dependencies" %}
+
+3. Click the `+` button.
```
```suggestion
3. Click the **+** button.
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+   caption:"The project's package dependencies" %}
+
+3. Click the `+` button.
+4. Click the `Add Local...` button on the bottom of the dialog that opens.
```
```suggestion
4. In the dialog that opens, click **Add Local...**.
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+3. Click the `+` button.
+4. Click the `Add Local...` button on the bottom of the dialog that opens.
+5. Navigate to `ios/Flutter/ephemeral/Packages/FlutterGeneratedPluginSwiftPackage`
+   and click the `Add Package` button.
```
```suggestion
   and click the **Add Package** button.
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+4. Click the `Add Local...` button on the bottom of the dialog that opens.
+5. Navigate to `ios/Flutter/ephemeral/Packages/FlutterGeneratedPluginSwiftPackage`
+   and click the `Add Package` button.
+6. Ensure it is added to the `Runner` target and click the `Add Package` button.
```
```suggestion
6. Ensure that it's added to the **Runner** target and click the `Add Package` button.
```

(The general rule of thumb, is that menu items should be bolded.)

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+   image:"development/packages-and-plugins/swift-package-manager/choose-package-products.png",
+   caption:"Ensure the package is added to the `Runner` target" %}
+
+7. Ensure `FlutterGeneratedPluginSwiftPackage` was added to Frameworks,
+   Libraries, and Embedded Content.
```
```suggestion
7. Ensure that `FlutterGeneratedPluginSwiftPackage` was added to **Frameworks,
   Libraries, and Embedded Content**.
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+
+### Step 2: Add Run Prepare Flutter Framework Script Pre-Action {:.no_toc}
+
+**The following must be completed for each flavor.**
```
```suggestion
**The following steps must be completed for each flavor.**
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+
+**The following must be completed for each flavor.**
+
+1. Next, select Product > Scheme > Edit Scheme.
+2. Click the `>` next to "Build" in the left side bar.
+3. Select Pre-actions.
+4. Click the `+` button and select `New Run Script Action` from the menu.
+5. Click the "Run Script" title and change to
+   `Run Prepare Flutter Framework Script`.
+6. Change the "Provide build settings from" to the app.
+7. Input the following in the text box:
```
```suggestion
1. Select **Product > Scheme > Edit Scheme**.
2. Click the **>** next to **Build** in the left side bar.
3. Select **Pre-actions**.
4. Click the **+** button and select **New Run Script Action** from the menu.
5. Click the **Run Script** title and change to
   **Run Prepare Flutter Framework Script**.
6. Change the **Provide build settings from** to the app.
7. Input the following in the text box:
```

loic-sharma replied:
About this bit:

> Click the **Run Script** title and change to
   **Run Prepare Flutter Framework Script**.

The "Run Prepare Flutter Framework Script" is text the user inputs into a text field (and not a menu item). I assume that should that be wrapped in backticks?

loic-sharma replied:
Fixed everything else!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+
+### Step 3: Run app {:.no_toc}
+
+1. Run the app in Xcode and ensure `FlutterGeneratedPluginSwiftPackage` is a
```
```suggestion
1. Run the app in Xcode and ensure that **FlutterGeneratedPluginSwiftPackage** is a
```

loic-sharma replied:
`FlutterGeneratedPluginSwiftPackage` is a Swift package identifier that is displayed in something that looks like menu text... I believe backticks are preferable here, but let me know if I'm incorrect!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+### Step 3: Run app {:.no_toc}
+
+1. Run the app in Xcode and ensure `FlutterGeneratedPluginSwiftPackage` is a
+   target dependency and `Run Prepare Flutter Framework Script` is ran as
```
```suggestion
   target dependency and **Run Prepare Flutter Framework Script** runs as
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+   image:"development/packages-and-plugins/swift-package-manager/flutter-pre-action-build-log.png",
+   caption:"Ensure `Run Prepare Flutter Framework Script` is ran as a pre-action" %}
+
+2. Also, ensure the app runs on the command line with `flutter run`.
```
```suggestion
2. Also, ensure that the app runs on the command line with `flutter run`.
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-macos-project-manually.md:
```diff
@@ -0,0 +1,74 @@
+Once you enable the Swift Package Manager feature, the Flutter CLI will try
```
```suggestion
Once you enable Swift Package Manager support, the Flutter CLI tries
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-macos-project-manually.md:
```diff
@@ -0,0 +1,74 @@
+to migrate your project to add Swift Package Manager integration. This makes
+your project download the Swift packages that your Flutter plugins depend on.
+
+However, the Flutter CLI may be unable to migrate your project if you've
```
```suggestion
However, the Flutter CLI might be unable to migrate your project if you've
```

You might wonder why I keep changing "may" to "might". It's in our wordlist. "May" is generally used in legal contexts.  https://developers.google.com/style/word-list#may

loic-sharma replied:
Thanks for the link, that's a good resource! :)

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-macos-project-manually.md:
```diff
@@ -0,0 +1,74 @@
+However, the Flutter CLI may be unable to migrate your project if you've
+modified it significantly.
+
+Please [file a bug][] before migrating manually to help the Flutter team improve
+the automatic migration.
+Please include the error message you received and consider including a copy of
+the of the following files in your bug report:
```
Same edits as above.

loic-sharma replied:
Updated!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-macos-project-manually.md:
```diff
@@ -0,0 +1,74 @@
+### Step 1: Add FlutterGeneratedPluginSwiftPackage Package Dependency {:.no_toc}
+
+1. Open your app (`macos/Runner.xcworkspace`) in Xcode.
+2. Navigate to Package Dependencies for the project.
```
Same edits as above. (Menu text should be bolded, for example.)

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-macos-project-manually.md:
```diff
@@ -0,0 +1,74 @@
+   image:"development/packages-and-plugins/swift-package-manager/package-dependencies.png",
+   caption:"The project's package dependencies" %}
+
+3. Click the `+` button.
```
Same edits as above...

loic-sharma replied:
Fixed!

sfshaza2 reviewed (commented):
Hi, @loic-sharma! Thanks so much for this PR. I have reviewed a chunk of it, but some pages have the same text (and therefore require the same edits). Let me know when you are ready for me to take another look.

loic-sharma commented on src/content/packages-and-plugins/swift-package-manager/for-plugin-authors.md:
```diff
@@ -0,0 +1,141 @@
+notice.
+
+Swift Package Manager adoption will be gradual.
+Plugins that don't support CocoaPods won't be usable by projects that haven't
+migrated to Swift Package Manager yet.
+Plugins that don't support Swift Package Manager can cause problems for projects
+that have migrated.
+
+This guide shows how to add Swift Package Manager support to a plugin that
+already supports CocoaPods.
+This ensures the plugin is usable by all Flutter projects.
```
@sfshaza2 FYI, this content is new since your last review

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
````diff
@@ -0,0 +1,62 @@
+1. Switch to Flutter's `master` channel:
+
+   ```sh
+   flutter channel master
````
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,62 @@
+Flutter's Swift Package Manager feature is disabled by default.
+To enable it:
+
+1. Switch to Flutter's `master` channel:
```
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
````diff
@@ -0,0 +1,62 @@
+   flutter config --enable-swift-package-manager
+   ```
+
+Running an app using the Flutter CLI will migrate it to add Swift Package
````
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,62 @@
+dependencies.
+However, Swift Package Manager will remain integrated with your project.
+To remove this integration,
+follow the ["How to remove Swift Package Manager integration"][] instructions.
```
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,62 @@
+
+### Disable for a single project
+
+In the project's pubspec.yaml, under the `flutter` section,
```
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
````diff
@@ -0,0 +1,62 @@
+flutter config --no-enable-swift-package-manager
+```
+
+["How to remove Swift Package Manager integration"]: /packages-and-plugins/swift-package-manager/for-app-developers#how-to-remove-swift-package-manager-integration
````
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+to migrate your project to add Swift Package Manager integration. This makes
+your project download the Swift packages that your Flutter plugins depend on.
+
+However, the Flutter CLI may be unable to migrate your project if you've
```
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,62 @@
+follow the ["How to remove Swift Package Manager integration"][] instructions.
+
+:::tip
+If you find a bug in Flutter's Swift Package Manager feature,
```
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,76 @@
+Once you enable the Swift Package Manager feature, the Flutter CLI will try
```
Fixed!

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,62 @@
+This makes your project download the Swift packages that
+your Flutter plugins depend on.
+
+:::note
+Flutter will fallback to CocoaPods for dependencies that do not support Swift
+Package Manager yet.
+:::
```
This isn't quite right. A Flutter app can have a dependencies from both CocoaPods and SwiftPM at the same time. However, mixing dependencies from both package managers can cause issues so ideally everyone adds SwiftPM support ASAP.

loic-sharma replied:
Additional context: an app could have a dependency that only supports SwiftPM and another dependency that only supports CocoaPods. We _have_ to use both package managers in this scenario.

Arguably it'd be better if we tried to stick to just one package manager if possible as described by your suggestion... but we don't do that today.

sfshaza2 replied:
I defer to you on the wording...

loic-sharma commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,62 @@
+:::
+
+## How to disable Swift Package Manager
+
```
Thanks for the suggestion! I moved the link to file a bug to this box to reduce the number of callouts in this section. Let me know if you have concerns with that

loic-sharma commented:
@sfshaza2 Thanks for the thorough feedback, this is ready for another round of review!

loic-sharma commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
```diff
@@ -0,0 +1,344 @@
+   Name this new directory the name of the platform package.
+
+   <pre>
+   /plugin_name/plugin_name_ios/ios/<b>plugin_name_ios</b>
```
Note: a customer might not have the `plugin_name/plugin_name_ios/` directory structure. This is a commonly used pattern for federated plugins, but a user could also have a directory structure like `/plugin_name/ios/...`. Please let me know if you have suggestions to improve clarity.

loic-sharma commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
```diff
@@ -0,0 +1,344 @@
+      - This is needed to ensure the directory is committed, even if empty.
+        Can be removed if files are added to the directory.
+
+   <pre>
+   /plugin_name/plugin_name_ios/ios/plugin_name_ios/<b>Package.swift</b>
+   /plugin_name/plugin_name_ios/ios/plugin_name_ios/<b>Sources</b>
+   /plugin_name/plugin_name_ios/ios/plugin_name_ios/<b>Sources/plugin_name_ios</b>
+   /plugin_name/plugin_name_ios/ios/plugin_name_ios/<b>Sources/plugin_name_ios/include</b>
+   /plugin_name/plugin_name_ios/ios/plugin_name_ios/<b>Sources/plugin_name_ios/include/plugin_name_ios</b>
+   /plugin_name/plugin_name_ios/ios/plugin_name_ios/<b>Sources/plugin_name_ios/include/plugin_name_ios/.gitkeep</b>
+   </pre>
```
Maybe directory ASCII art would be better here? Something like:

```suggestion
   <pre>
   /plugin_name/plugin_name_ios/ios/
   \- plugin_name_ios/
      \- <b>Package.swift</b>
      \- <b>Sources/plugin_name_ios/include/plugin_name_ios</b>
         \- <b>.gitkeep</b>
   </pre>
```

loic-sharma replied:
I switched to this!

loic-sharma commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
```diff
@@ -0,0 +1,365 @@
+    @@ -21,4 +21,4 @@ 
+    - s.source_files = 'Classes/**/*.{h,m}'
+    - s.public_header_files = 'Classes/**/*.h'
+    - s.module_map = 'Classes/cocoapods_plugin_name_ios.modulemap'
```
This `podspec` contains attributes that customers might not have.

For example, [only ~70 out of ~10K plugins have a `modulemap`](https://gist.github.com/loic-sharma/891a358f8abab36133e8a9395f719296). Folks that don't have a `modulemap` don't need to add/change a `module_map` attribute.

Hopefully that's clear, but let me know if you have recommendations to improve clarity here.

sfshaza2 commented:
@loic-sharma, please ping me directly when you are ready for me to review again. THANKS!!

jmagman commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,129 @@
+
+Flutter is migrating to [Swift Package Manager][] to manage iOS and macOS native
+dependencies.
+This is an experimental feature that might change in the future.
+It is currently only available on the [`main` channel][].
+Flutter will continue to support CocoaPods until further notice.
```
Can we hoist this into an `:::important` banner at the top, or whatever strong indicator the rest of the site uses for experimental docs?

loic-sharma replied:
I wrapped this in a `:::important` banner. @sfshaza2 Let me know if there's a better way to label experimental docs.

sfshaza2 replied:
All you need to do is:

:::important
Your content here
:::

loic-sharma replied:
Done! :)

jmagman commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,67 @@
+
+### Disable globally for all projects
+
+Run the following command:
```
Not sure if this is worth pointing out, but it would need to be disabled for _all_ contributors to an app as well?

loic-sharma replied:
Added! I'll leave this open for now in case folks have suggestions on the wording.

jmagman commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
```diff
@@ -0,0 +1,372 @@
+   version of the plugin name.
+   :::
+
+5. If your plugin has a `PrivacyInfo.xcprivacy` file, move it to
```
Link to Apple docs https://developer.apple.com/documentation/bundleresources/privacy_manifest_files ?

loic-sharma replied:
Good idea, added!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
```diff
@@ -0,0 +1,377 @@
+               name: "plugin_name",
+               dependencies: [],
+               resources: [
+                   // If your plugin requires a privacy manifest, for example if it uses any required
```
Required what? It feels like this is chopped off.

loic-sharma replied:
`Required reason APIs` is what Apple calls these: https://developer.apple.com/documentation/bundleresources/privacy_manifest_files/describing_use_of_required_reason_api

I'll re-wrap these lines to make it clearer this is an intentional phrase and not a mistake :)

loic-sharma replied:
Re-wrapped! I'll leave this open in case folks have feedback

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
````diff
@@ -0,0 +1,377 @@
+
+   ```swift title="Package.swift"
+               resources: [
+                   // If your plugin requires a privacy manifest, for example if it uses any required
````
Ditto.

loic-sharma replied:
See: https://github.com/flutter/website/pull/10827#discussion_r1674814361

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
```diff
@@ -0,0 +1,377 @@
+
+6. Move any resource files from `ios/Assets` to `ios/Sources/plugin_name`
+   (or a subdirectory).
+   Add the resource files to your `Package.swift` file if applicable.
```
```suggestion
   Add the resource files to your `Package.swift` file, if applicable.
```

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
```diff
@@ -0,0 +1,377 @@
+    * If you're unsure which headers are public, check your `podspec` file's
+      [`public_header_files`][] attribute.
+      If this attribute  is not specified, all of your headers were public.
+      You should consider whether or not you want all of your headers to be
```
```suggestion
      You should consider whether you want all of your headers to be
```
https://developers.google.com/style/word-list#whether

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-objective-c-plugin.md:
```diff
@@ -0,0 +1,377 @@
+
+   If you're using a `modulemap` for CocoaPods to create a Test submodule,
+   consider removing it for Swift Package Manager.
+   Note that this will make all public headers available via the module.
```
```suggestion
   Note that this makes all public headers available through the module.
```

https://developers.google.com/style/word-list#via

loic-sharma replied:
Fixed!

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-swift-plugin.md:
```diff
@@ -0,0 +1,242 @@
+               name: "plugin_name",
+               dependencies: [],
+               resources: [
+                   // If your plugin requires a privacy manifest, for example if it uses any required
```
Again, "required reason APIs"?

loic-sharma replied:
See: https://github.com/flutter/website/pull/10827#discussion_r1674814361

sfshaza2 commented on src/_includes/docs/swift-package-manager/migrate-swift-plugin.md:
```diff
@@ -0,0 +1,242 @@
+
+    2. Run `flutter run` with the example app and ensure it builds and runs.
+
+    3. Open the example app in Xcode and ensure **Package Dependencies** show in
```
```suggestion
    3. Open the example app in Xcode and ensure that **Package Dependencies** shows in
```

loic-sharma replied:
Fixed!

sfshaza2 reviewed (approved):
@loic-sharma, what a beast! I think it's pretty clear though. I had some very minor feedback this time, so I'm going to approve it, to save us both time. Assuming your tech reviewers are happy, you can land this after a few minor tweaks. THANKS

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+---
+
+:::warning
+Flutter is migrating to [Swift Package Manager][] to manage iOS and macOS native
+dependencies.
+This is an experimental feature that might change in the future.
+It is currently only available on the [`main` channel][].
+Flutter will continue to support CocoaPods until further notice.
```
**issue:** Condition, then instruction per [GSG](https://developers.google.com/style/sentence-structure).
Keep the documentation timeless per [GSG](https://developers.google.com/style/timeless-documentation).

```suggestion
To manage iOS and macOS native dependencies,
Flutter is migrating to [Swift Package Manager][] .
Flutter's support of SPM as described in this guide is under development. 
The implementation might change in the future.
SPM support can be found on the [`main` channel][] only.
Flutter continues to support CocoaPods.
```

loic-sharma replied:
Normally I agree that "condition, then instruction" is better, but I think it makes the first sentence choppy. I personally prefer the original first sentence. How strongly do you feel about this?

I updated the remaining sentences!

jmagman commented on src/_data/sidenav.yml:
```diff
@@ -502,6 +502,13 @@
+    - title: Swift Package Manager
+      permalink: /packages-and-plugins/swift-package-manager
+      children:
+        - title: For app developers
+          permalink: /packages-and-plugins/swift-package-manager/for-app-developers
```
I'm not sure app developers would think to go into `Packages & plugins`
Maybe this should move into `Platform integration`? Though it would make sense under iOS and macOS, and we don't have a shared category.  Whatever you and the tech writers think works best.

sethladd replied:
oh, good idea!

loic-sharma replied:
Sorry for the slow response, I'd like to test if I can link to this page twice: once from the packages & plugin section and once from the platform integration section. Will loop back soon!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+   macOS.
+
+If you find a bug in Flutter's Swift Package Manager support,
+please [open an issue][].
```
**issue:** Remove _please_ per [GSG](https://developers.google.com/style/tone#politeness-and-use-of-please).

```suggestion
[open an issue][].
```

loic-sharma replied:
This feels curt to me, but updated :(

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+
+## How to add Flutter to an existing app (add-to-app)
+
+Flutter's Swift Package Manager support doesn't work with add-to-app scenarios
+yet.
+
+For status updates, see [flutter#146957].
```
**issue:** Keep the documentation timeless per [GSG](https://developers.google.com/style/timeless-documentation).

```suggestion
Flutter's Swift Package Manager support doesn't work with add-to-app scenarios.

To keep current on status updates, consult [flutter#146957].
```

loic-sharma replied:
Updated!

jmagman reviewed (approved):
Thank you so much for tackling this!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
````diff
@@ -0,0 +1,144 @@
+Target Integrity (Xcode): The package product 'plugin_name_ios' requires minimum platform version 14.0 for the iOS platform, but this target supports 12.0
+```
+
+To use the plugin, you'll need to increase the **Minimum Deployments** of your
+app's target.
````
**issue:** Use present tense per [GSG](https://developers.google.com/style/tense).

```suggestion
To use the plugin, increase the **Minimum Deployments** of your app's target.
```

loic-sharma replied:
Nice, updated!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+To use the plugin, you'll need to increase the **Minimum Deployments** of your
+app's target.
+
+:::warning
+This raises your app's minimum OS version requirement.
+:::
```
**suggestion:** This is probably redundant.

```suggestion
```

loic-sharma replied:
That's fair, removed!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+
+## How to add Swift Package Manager integration to a custom Xcode target
+
+Your Flutter Xcode project can have custom [Xcode targets][] to build additional
+products, like frameworks or unit tests.
+You can add Swift Package Manager integration to these custom Xcode targets
+as well.
```
**issue:** Condition, then instruction per [GSG](https://developers.google.com/style/sentence-structure).

```suggestion
To build additional products like frameworks or unit tests,
your Flutter Xcode project can have custom [Xcode targets][].
You can add Swift Package Manager integration to these custom Xcode targets.
```

loic-sharma replied:
This first sentence isn't an instruction or a condition, it's more of a background "here's what custom Xcode targets are used for" in case someone is wondering if this section is for them. I personally prefer the first sentence as-is, how strongly do you feel about this?

I updated the second sentence!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+
+## How to remove Swift Package Manager integration
+
+The Flutter CLI migrates your project to add Swift Package Manager integration.
+This makes your project download the Swift packages that your Flutter plugins
+depend on.
```
**issue:** 
Condition, then instruction per [GSG](https://developers.google.com/style/sentence-structure).

```suggestion
To add Swift Package Manager integration, the Flutter CLI migrates your project.
During this migration, your project downloads the Swift packages
on which your Flutter plugins depend.
```

loic-sharma replied:
Mm my second sentence was incorrect. Sorry about that.

I updated this to:

> To add Swift Package Manager integration, the Flutter CLI migrates your project.
> This migration updates your Xcode project to add Flutter plugin dependencies.

Let me know what you think!

sethladd commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
````diff
@@ -0,0 +1,77 @@
+   flutter config --enable-swift-package-manager
+   ```
+
+Running an app using the Flutter CLI migrates it to add Swift Package Manager
````
nit: this sentence was a bit hard to follow.

every time I run the app, the app is migrated?
what is "it" in the above? the CLI or my app? (I assume the app, but would be good to spell out)

loic-sharma replied:
Good feedback!

> every time I run the app, the app is migrated?

Yup that's correct. Each time you build your app, the tool will check if it needs to modify the app.

loic-sharma replied:
Updated to the suggestion here: https://github.com/flutter/website/pull/10827#discussion_r1678456670

sethladd commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,77 @@
+:::secondary Plugin authors
+Plugin authors need to turn on and off Flutter's Swift Package Manager
+support for testing.
+Otherwise, you likely won't need to turn this off.
```
maybe: App developers do not need to disable Swift Package Manager support, unless they are running into bugs or issues.

loic-sharma replied:
Nice, updated!

sethladd commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
```diff
@@ -0,0 +1,77 @@
+Disabling Swift Package Manager causes Flutter to use CocoaPods for all
+dependencies.
+However, Swift Package Manager remains integrated with your project.
+To remove this integration,
```
maybe: "To remove Swift Package Manager completely from your app project,"

disambiguating "this"

loic-sharma replied:
Updated!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+
+3. Navigate to **Package Dependencies** for the project.
+
+4. Click on the `FlutterGeneratedPluginSwiftPackage` package and then click the
+   **-** button.
```
**issue:** Use the button's label per [GSG](https://developers.google.com/style/ui-elements#buttons).

```suggestion
4. Click the `FlutterGeneratedPluginSwiftPackage` package,
    then click <span class="material-symbols-outlined">add</span>.
```

loic-sharma replied:
Updated. FYI the `-` appears too high:
 
![image](https://github.com/user-attachments/assets/52e332ab-4e42-4bea-b808-c3551f4fa1f5)

It looks like I need to add a `vertical-align: text-bottom` to the `span`. 

![image](https://github.com/user-attachments/assets/95b86de4-3470-4be6-8b90-074ce93e6644)

Should I add that globally on the website for all `p span.material-symbols-outlined`?

sethladd commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,84 @@
+Once you [turn on Swift Package Manager][] support, the Flutter CLI tries to
+migrate your project to add Swift Package Manager integration.
+This makes your project download the Swift packages that your Flutter
```
maybe a bit more clear to say "the Flutter tooling downloads the Swift packages" ?

loic-sharma replied:
Swift packages are downloaded by Xcode. The Flutter tool just generates some glue.

The process is roughly:

1. The Flutter tool download pub packages
2. The Flutter tool generates code to add iOS/macOS plugins to the Xcode project
3. Xcode downloads the Swift package dependencies for each iOS/macOS plugin

Before SwiftPM, the generated code in step 2 was CocoaPods magic. Now, the generated code also includes SwiftPM magic. The "Swift Package Manager integration" is your project including the SwiftPM magic that was generated by step 2.

sethladd commented on src/_includes/docs/swift-package-manager/migrate-ios-project-manually.md:
```diff
@@ -0,0 +1,84 @@
+This makes your project download the Swift packages that your Flutter
+plugins depend on.
+
+However, the Flutter CLI might be unable to migrate your project if you've
```
Do we mean "even if you've modified it significantly" ?

Note: "even" here?

loic-sharma replied:
We can migrate projects _unless_ they've received heavy modifications.

We match do "match and replace" in your Xcode project to migrate it. This should work fine if your project is based off Flutter's templates. However, our matching logic can break if you've modified your Xcode project. The bigger your modifications, the more likely our matching logic breaks.

sethladd replied:
I guess I'm wondering if we need the "significantly" qualifier?

ideas:

"The Flutter CLI attempts to migrate your project to Swift Package Manager, even if you have made changes to <foo>. If the Flutter CLI tool is unable to migrate your project (for example, if there are unexpected modifications to Foo or Bar), then the Flutter CLI tool will <print some error or warning telling you where to manually update and modify>."

loic-sharma replied:
Ah gotcha, I dropped the "significantly" qualifier and switched to wording similar to what you suggested.

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+4. Navigate to **Frameworks, Libraries, and Embedded Content** for the `Runner`
+   target.
+
+5. Click on `FlutterGeneratedPluginSwiftPackage` and then click the **-**
+   button.
```
**issue:** Use _click_, not _click on_ per [GSG](https://developers.google.com/style/word-list#click).

```suggestion
5. Click `FlutterGeneratedPluginSwiftPackage`,
   then click <span class="material-symbols-outlined">remove</span>.
```

loic-sharma replied:
Updated!

sethladd commented on src/_includes/docs/swift-package-manager/migrate-macos-project-manually.md:
```diff
@@ -0,0 +1,84 @@
+Once you [turn on Swift Package Manager][] support, the Flutter CLI tries to
+migrate your project to add Swift Package Manager integration.
```
maybe: "the next time the CLI is run"

sethladd commented on src/_includes/docs/swift-package-manager/migrate-macos-project-manually.md:
```diff
@@ -0,0 +1,84 @@
+This makes your project download the Swift packages that your Flutter
+plugins depend on.
+
+However, the Flutter CLI might be unable to migrate your project if you've
```
same: "even if" , here, right?

loic-sharma replied:
I dropped the "significantly" qualifier and switched to wording similar to what you suggested here: https://github.com/flutter/website/pull/10827#discussion_r1680007985

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+   image:"development/packages-and-plugins/swift-package-manager/remove-generated-framework.png",
+   caption:"The `FlutterGeneratedPluginSwiftPackage` to remove" %}
+
+6. Select **Product > Scheme > Edit Scheme**.
```
**suggestion:** Phrase as _Go to_ instead of _Select_ per [GSG](https://developers.google.com/style/ui-elements#ui-terms).

```suggestion
6. Go to **Product > Scheme > Edit Scheme**.
```

loic-sharma replied:
Updated!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+
+6. Select **Product > Scheme > Edit Scheme**.
+
+7. Click the **>** next to **Build** in the left side bar.
```
**suggestion:** Focus on the task per [GSG](https://developers.google.com/style/ui-elements#task-focus).

```suggestion
7. Expand the **Build** section in the left side bar.
```

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+
+7. Click the **>** next to **Build** in the left side bar.
+
+8. Select **Pre-actions**.
+
+9. Select the **Run Prepare Flutter Framework Script**.
```
**issue:** Use _click_ and _expand_.

```suggestion
8. Click **Pre-actions**.

9. Expand **Run Prepare Flutter Framework Script**.
```

loic-sharma replied:
Updated!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+
+9. Select the **Run Prepare Flutter Framework Script**.
+
+10. Click the **<span class="material-symbols">delete</span>** button.
```
**issue:** Use _click_.

```suggestion
10. Click **<span class="material-symbols">delete</span>**.
```

loic-sharma replied:
Updated!

atsansone commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,144 @@
+
+Flutter's Swift Package Manager integration has several benefits:
+
+1. **Access to the Swift package ecosystem**.
+   Flutter plugins can use the growing ecosystem of [Swift packages][]! 
+2. **Simplifies Flutter installation**.
+   Swift Package Manager is bundled with Xcode.
+   In the future, you won’t need to install Ruby and CocoaPods to target iOS or
+   macOS.
```
**issue:** 
- Use parallel structure in your list sentences.
- Remove the exclamation point per [GSG](https://developers.google.com/style/periods#exclamation-points).
- Keep the documentation timeless.
- Use present tense.

```suggestion
1. **Provides access to the Swift package ecosystem**.
   Flutter plugins can use the growing ecosystem of [Swift packages][]! 
1. **Simplifies Flutter installation**.
   Xcode includes Swift Package Manager.
   Targeting iOS or macOS doesn't require Ruby and CocoaPods.
```

loic-sharma replied:
This sounds good but I tweaked the last sentence:

```diff
- Targeting iOS or macOS doesn't require Ruby and CocoaPods.
+ You don't need to install Ruby and CocoaPods if your project uses
+ Swift Package Manager.
```

Let me know what you think!

atsansone reviewed (commented):
@loic-sharma : I suggested changes to one file. Review and accept/decline and apply those you accept to the other pages. Thanks!

stuartmorgan commented on src/content/packages-and-plugins/swift-package-manager/for-plugin-authors.md:
```diff
@@ -0,0 +1,155 @@
+Plugins that don't support CocoaPods won't be usable by projects that haven't
+migrated to Swift Package Manager yet.
+Plugins that don't support Swift Package Manager can cause problems for projects
+that have migrated.
```
It might be nice to explicitly say here that we expect to remove support for CocoaPods at some point in the future, so adding SPM support will future-proof plugins. That will make it more clear why a plugin author should do this work.

loic-sharma replied:
Agreed.

@atsansone What would be the best way to state this while following the timeless [GSG](https://developers.google.com/style/timeless-documentation)?

> Flutter expects to remove support for CocoaPods. Adding Swift Package Manager support future-proofs your plugins.

I think this would be a valuable sentence to add, but it goes against the GSG guidance.

sethladd commented on src/_includes/docs/swift-package-manager/how-to-enable-disable.md:
````diff
@@ -0,0 +1,78 @@
+   flutter config --enable-swift-package-manager
+   ```
+
+Using the Flutter CLI to run an app modifies the app to add Swift Package
````
nit: does it modify the app, or modify the project?

loic-sharma replied:
Technically both is true, but project is more precise. Updated!

sethladd commented on src/content/packages-and-plugins/swift-package-manager/for-app-developers.md:
```diff
@@ -0,0 +1,138 @@
+
+{% include docs/swift-package-manager/how-to-enable-disable.md %}
+
+## How to add Flutter to an existing app (add-to-app)
```
this was a surprising header in this file, and it was surprising to see this first before "how to add SPM to a project".

Consider:

* How to add SPM to a full Flutter app

<existing doc below>

* How to migrate to SPM for a Flutter add-to-app scenario

Not yet supported, follow this bug, etc...


In this way, the headers rhyme and are clear about their scenarios. WDYT?

loic-sharma replied:
This is excellent feedback. The guide was misleading: hopefully no one needs to do the manual steps. I restructured the page and added a guide dedicated to the auto migration. I also renamed these headers to align better. See: https://github.com/flutter/website/pull/10827#issuecomment-2231753344

johnpryan reviewed (approved):
LGTM - i was able to get the `pedometer` sample up and running with these instructions.

loic-sharma commented:
@sfshaza2 @atsansone @sethladd Hey there, I re-structured the "For app developers page":

![Screenshot 2024-07-16 at 1 09 20 PM](https://github.com/user-attachments/assets/115c9aed-75ae-49a7-8fa8-7f25bdf143af)

Summary of changes:

1. I added a new **How to add Swift Package Manager Integration** section. I moved other existing guides to this section (the add-to-app guide, and the manual process guide)
2. I created a new guide **How to add Swift Package Manager Integration** > **Add to a Flutter app*. This section is entirely new content

This was in response to Seth's comment here: https://github.com/flutter/website/pull/10827#discussion_r1679659417

Please let me know what you think!

sethladd reviewed (approved):
lgtm - thank you!

sfshaza2 commented:
@loic-sharma, thanks for all of these updates and hanging in there! I'm going to land this. If there are further tweaks, they can be handled in another PR.
